<launch>
  <arg name="show_debug_info" default="false"/>
  <arg name="enable_yaw_bias_estimation" default="True"/>
  <arg name="predict_frequency" default="50.0"/>
  <arg name="tf_rate" default="10.0"/>
  <arg name="extend_state_step" default="50"/>

  <arg name="input_initial_pose_name" default="initialpose3d"/>
  <arg name="input_trigger_node_service_name" default="trigger_node" description="trigger node service"/>

  <!-- for Pose measurement -->
  <arg name="input_pose_with_cov_name" default="in_pose_with_covariance"/>
  <arg name="pose_additional_delay" default="0.0"/>
  <arg name="pose_measure_uncertainty_time" default="0.01"/>
  <arg name="pose_smoothing_steps" default="5"/>
  <arg name="pose_gate_dist" default="10000.0"/>

  <!-- for twist measurement -->
  <arg name="input_twist_with_cov_name" default="in_twist_with_covariance"/>
  <arg name="twist_additional_delay" default="0.0"/>
  <arg name="twist_smoothing_steps" default="2"/>
  <arg name="twist_gate_dist" default="10000.0"/>

  <!-- for process model -->
  <arg name="proc_stddev_yaw_c" default="0.005"/>
  <arg name="proc_stddev_vx_c" default="5.0"/>
  <arg name="proc_stddev_wz_c" default="1.0"/>

  <!-- output topic name -->
  <arg name="output_odom_name" default="ekf_odom"/>
  <arg name="output_pose_name" default="ekf_pose"/>
  <arg name="output_pose_with_covariance_name" default="ekf_pose_with_covariance"/>
  <arg name="output_biased_pose_name" default="ekf_biased_pose"/>
  <arg name="output_biased_pose_with_covariance_name" default="ekf_biased_pose_with_covariance"/>
  <arg name="output_twist_name" default="ekf_twist"/>
  <arg name="output_twist_with_covariance_name" default="ekf_twist_with_covariance"/>

  <node pkg="ekf_localizer" exec="ekf_localizer" name="ekf_localizer" output="screen">
    <remap from="in_pose_with_covariance" to="$(var input_pose_with_cov_name)"/>

    <remap from="in_twist_with_covariance" to="$(var input_twist_with_cov_name)"/>

    <remap from="initialpose" to="$(var input_initial_pose_name)"/>
    <remap from="trigger_node_srv" to="$(var input_trigger_node_service_name)"/>

    <param name="pose_frame_id" value="map"/>

    <param name="show_debug_info" value="$(var show_debug_info)"/>
    <param name="enable_yaw_bias_estimation" value="$(var enable_yaw_bias_estimation)"/>

    <param name="predict_frequency" value="$(var predict_frequency)"/>
    <param name="tf_rate" value="$(var tf_rate)"/>
    <param name="extend_state_step" value="$(var extend_state_step)"/>

    <param name="pose_additional_delay" value="$(var pose_additional_delay)"/>
    <param name="pose_measure_uncertainty_time" value="$(var pose_measure_uncertainty_time)"/>
    <param name="pose_smoothing_steps" value="$(var pose_smoothing_steps)"/>
    <param name="pose_gate_dist" value="$(var pose_gate_dist)"/>

    <param name="twist_additional_delay" value="$(var twist_additional_delay)"/>
    <param name="twist_smoothing_steps" value="$(var twist_smoothing_steps)"/>
    <param name="twist_gate_dist" value="$(var twist_gate_dist)"/>

    <param name="proc_stddev_yaw_c" value="$(var proc_stddev_yaw_c)"/>
    <param name="proc_stddev_vx_c" value="$(var proc_stddev_vx_c)"/>
    <param name="proc_stddev_wz_c" value="$(var proc_stddev_wz_c)"/>

    <remap from="ekf_odom" to="$(var output_odom_name)"/>
    <remap from="ekf_pose" to="$(var output_pose_name)"/>
    <remap from="ekf_pose_with_covariance" to="$(var output_pose_with_covariance_name)"/>
    <remap from="ekf_biased_pose" to="$(var output_biased_pose_name)"/>
    <remap from="ekf_biased_pose_with_covariance" to="$(var output_biased_pose_with_covariance_name)"/>
    <remap from="ekf_twist" to="$(var output_twist_name)"/>
    <remap from="ekf_twist_with_covariance" to="$(var output_twist_with_covariance_name)"/>
  </node>
</launch>
